pkg/engine/lifecycletest/refresh_test.go

```diff
@@ -1844,3 +1846,119 @@ func TestRefreshWithProgramWithDeletedResource(t *testing.T) {
 	// Should only have 2 resources now, the deleted one should be gone.
 	assert.Len(t, snap.Resources, 2)
 }
+
+// Regression test for https://github.com/pulumi/pulumi/issues/19406. Check that if we have a program with
+// more resources than step executor workers we don't deadlock.
```

Comment:
Don't suppose there are risks of this happening with other pump workflows like parallel diff/program destroy?

Reply:
Yeh but the same fix should apply